jose-util/generate.go

```diff
@@ -122,26 +122,42 @@ func GenerateEncryptionKey(alg jose.KeyAlgorithm, bits int) (crypto.PublicKey, c
 }
 
 func generate() {
-	kid := *generateKeyIdentFlag
-	if kid == "" {
-		kid = uuid.New().String()
-	}
-
-	var pubKey crypto.PrivateKey
-	var privKey crypto.PublicKey
+	var privKey crypto.PrivateKey
+	var pubKey crypto.PublicKey
 	var err error
 
 	switch *generateUseFlag {
 	case "sig":
 		pubKey, privKey, err = GenerateSigningKey(jose.SignatureAlgorithm(*generateAlgFlag), *generateKeySizeFlag)
 	case "enc":
 		pubKey, privKey, err = GenerateEncryptionKey(jose.KeyAlgorithm(*generateAlgFlag), *generateKeySizeFlag)
+	default:
+		// According to RFC 7517 section-8.2.  This is unlikely to change in the
+		// near future. If it were, new values could be found in the registry under
+		// "JSON Web Key Use": https://www.iana.org/assignments/jose/jose.xhtml
+		app.FatalIfError(errors.New("invalid key use.  Must be \"sig\" or \"enc\""), "unable to generate key")
 	}
 	app.FatalIfError(err, "unable to generate key")
 
-	pub := jose.JSONWebKey{Key: pubKey, KeyID: kid, Algorithm: *generateAlgFlag, Use: *generateUseFlag}
+	kid := *generateKeyIdentFlag
+
 	priv := jose.JSONWebKey{Key: privKey, KeyID: kid, Algorithm: *generateAlgFlag, Use: *generateUseFlag}
 
+	// Generate a canonical kid based on RFC 7638
+	if kid == "" {
+		thumb, err := priv.Thumbprint(crypto.SHA256)
+		if err != nil {
+			app.FatalIfError(err, "unable to compute thumbprint")
```

Comment:
Nit: no need for the `if err !== nil` block here, the `FatalIfError` function will check if there was an error.